includes/Core/Email_Reporting/Email_Log.php

```diff
@@ -0,0 +1,340 @@
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_error_details' ),
+			)
+		);
+
+		register_post_meta(
+			self::POST_TYPE,
+			self::META_REPORT_REFERENCE_DATES,
+			array(
+				'type'              => 'string',
+				'single'            => true,
+				'auth_callback'     => $auth_callback,
+				'sanitize_callback' => array( __CLASS__, 'sanitize_reference_dates' ),
+			)
+		);
+	}
+
+	/**
+	 * Sanitizes the report frequency meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return string Sanitized value.
+	 */
+	public static function sanitize_frequency( $value ) {
+		$allowed = array(
+			Reporting_Settings::FREQUENCY_WEEKLY,
+			Reporting_Settings::FREQUENCY_MONTHLY,
+			Reporting_Settings::FREQUENCY_QUARTERLY,
+		);
+		$value   = is_string( $value ) ? strtolower( $value ) : '';
+
+		return in_array( $value, $allowed, true ) ? $value : '';
+	}
+
+	/**
+	 * Sanitizes the batch ID meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return string Sanitized value.
+	 */
+	public static function sanitize_batch_id( $value ) {
+		$value = sanitize_text_field( (string) $value );
+
+		return substr( $value, 0, 191 );
+	}
+
+	/**
+	 * Sanitizes the send attempts meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return int Sanitized value.
+	 */
+	public static function sanitize_attempts( $value ) {
+		if ( (int) $value < 0 ) {
+			return 0;
+		}
+
+		return absint( $value );
+	}
+
+	/**
+	 * Sanitizes the error details meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return string Sanitized value.
+	 */
+	public static function sanitize_error_details( $value ) {
+		if ( is_wp_error( $value ) ) {
+			$value = array(
+				'errors'     => $value->errors,
+				'error_data' => $value->error_data,
+			);
+		}
+
+		if ( is_array( $value ) || is_object( $value ) ) {
+			$encoded = wp_json_encode( $value, JSON_UNESCAPED_UNICODE );
+			return is_string( $encoded ) ? $encoded : '';
+		}
+
+		if ( is_string( $value ) ) {
+			// Treat existing JSON strings as-is by checking decode status instead of rebuilding them.
+			json_decode( $value, true );
+			if ( json_last_error() === JSON_ERROR_NONE ) {
+				return $value;
+			}
+
+			$encoded = wp_json_encode(
+				array(
+					'message' => $value,
+				),
+				JSON_UNESCAPED_UNICODE
+			);
+
+			return is_string( $encoded ) ? $encoded : '';
+		}
+
+		return '';
+	}
+
+	/**
+	 * Sanitizes the report reference dates meta value.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param mixed $value Meta value.
+	 * @return string Sanitized value.
+	 */
+	public static function sanitize_reference_dates( $value ) {
+		if ( ! is_array( $value ) && ! is_object( $value ) ) {
+			return '';
+		}
+
+		$keys = array( 'startDate', 'sendDate', 'compareStartDate', 'compareEndDate' );
+		$in   = (array) $value;
+		$out  = array_fill_keys( $keys, 0 );
```

Comment:
In/out is a bit less clear than `input`/`output`:
```suggestion
		$input   = (array) $value;
		$output  = array_fill_keys( $keys, 0 );
```

But even so, these names don't seem quite right; it seems like they're more "pre-sanitised" values.

Couldn't we do an array map here rather than creating the array in this sort of odd foreach loop? The `$in` variable here doesn't seem to serve a purpose other than as an intermediary, which makes me think an array map might be a better fit here. 🤔

Reply:
Added inline comment to explain why we are filling the `0` value to the keys above and renamed variables so it is more clear